File: RB/cpp/RiotBase/Variable/Number/OtWrapperNumber.cpp
#include "../../System/Rcast.h"
#include "OtWrapperNumber.h"

namespace RiotBase {

Rp *OtWrapperNumber_New( Rp *rp ) {
    return OtWrapperNumber::New( rp );
}

/**
  On a besoin de stocker la façon avec laquelle on est arrivés à la valeur en cours, d'où la nécessité de faire un reg_op dès le début.
  Rq: on stocke des Rp * sans correction de cur_xxx, il faudrait faire une méthode de correction
  Aussi à faire: optimisation des patches
*/
OtWrapperNumber::OtWrapperNumber( Rp *val, UsrId usr_id ) : OtWrapper<OtWrapperNumber>( usr_id ), val( val ) {
    if ( bool( this->val ) )
        reg_op( op_set_und{}, this->val, Number{} );
}

std::string OtWrapperNumber::type_name() {
    return "OtWrapperNumber";
}

const Rp *OtWrapperNumber::prox_ref() const {
    return val.rp;
}

void OtWrapperNumber::write_structure( std::ostream &os ) const {
    RO_CALL( write_structure, val.rp, os << type_name() << '(' ); os << ')';
}

void OtWrapperNumber::self_add( const Number &n ) {
    if ( this->right_to( op_add{}, {}, n ) ) {
        reg_op( op_add{}, n );
        val += n;
    }
}

void OtWrapperNumber::apply_op( op_set, Aod &aod, const Number &str ) {
    if ( val != str ) {
        val = str;
        this->sig_change();
    }
}

void OtWrapperNumber::apply_op( op_set_und, Aod &aod, const Number &str, const Number &old ) {
    apply_op( op_set{}, aod, str );
}

void OtWrapperNumber::apply_op( op_add, Aod &aod, const Number &str ) {
    if ( str ) {
        val += str;
        this->sig_change();
    }
}

bool OtWrapperNumber::right_to( op_set, UsrId as_usr, const Number &str ) {
    return str > val ? this->has_right( can_increase{}, as_usr ) : bool( str == val ) or this->has_right( can_decrease{}, as_usr );
}

bool OtWrapperNumber::right_to( op_set_und, UsrId as_usr, const Number &str, const Number &old ) {
    return right_to( op_set{}, as_usr, str );
}

bool OtWrapperNumber::right_to( op_add, UsrId as_usr, const Number &str ) {
    return str > 0 ? this->has_right( can_increase{}, as_usr ) : bool( str == 0 ) or this->has_right( can_decrease{}, as_usr );
}

void OtWrapperNumber::undo_patch( op_set, Aod &aod, Number &str ) {
    WARNING( "Undo on a non und version should not happen" );
}

void OtWrapperNumber::undo_patch( op_set_und, Aod &aod, Number &str, Number &old ) {
    val = old;
}

void OtWrapperNumber::undo_patch( op_add, Aod &aod, Number &str ) {
    val -= str;
}

void OtWrapperNumber::transmission( op_set_und, BBQ br, Number &str, Number &old ) {
    transmission( op_set{}, br, str );
}

void OtWrapperNumber::register_patch_op( op_set, BBQ bq, Number &n ) {
    register_patch_op( op_set_und{}, bq, n, val );
}

namespace {

RiotBase_DEF_FUNC_2( self_add_b, 1, tv0->isa<OtWrapperNumber>() and tv1->static_category() == Number::category ) {
    static_cast<OtWrapperNumber *>( v0 )->self_add( reinterpret_cast<const Number &>( v1 ) );
}
RiotBase_DEF_FUNC_2( self_sub_b, 1, tv0->isa<OtWrapperNumber>() and tv1->static_category() == Number::category ) {
    static_cast<OtWrapperNumber *>( v0 )->self_add( - reinterpret_cast<const Number &>( v1 ) );
}

}


} // namespace RiotBase

File: RB/ts/RiotBase/Variable/Number/OtWrapperNumber.ts
import OtWrapper from "../Ot/OtWrapper";
import Method    from "../Core/Method";

//
export default
class OtWrapperNumber extends OtWrapper {
    // patch types
    static op_set = {
        attrs: [ "Variable" ],
        right( rp: OtWrapperNumber, flags : number, nv : Variable ) : boolean {
            return nv.sup_immediate( rp.val ) ? OtWrapperNumber.can_increase.ok( flags ) : nv.equ_immediate( rp.val ) || OtWrapperNumber.can_increase.ok( flags );
        },
        valid( rp: OtWrapperNumber, nv : Variable ) : boolean {
            throw new Error( "TODO" );
        },
        apply_op( rp: OtWrapperNumber, aod, nv : Variable ) : void {
            Method.methods["self_set_b"].call_2( rp.val, nv );
            rp.sig_change();
        },
        undo( rp: OtWrapperNumber, aod, nv : Variable ) : void {
            throw new Error( "Weird" );
        },
        update_patch_data_l0( rp: OtWrapperNumber, args ) : void {
        }
    };
    static op_set_und = {
        attrs: [ "Variable", "Variable" ],
        right( rp: OtWrapperNumber, flags : number, val : Variable, old : Variable ) : boolean {
            return OtWrapperNumber.op_set.right( rp, flags, val );
        },
        valid( rp: OtWrapperNumber, val : Variable, old : Variable ) : boolean {
            throw new Error( "TODO" );
        },
        apply_op( rp: OtWrapperNumber, aod, val : Variable, old : Variable ) : void {
            OtWrapperNumber.op_set.apply_op( rp, aod, val );
        },
        undo( rp: OtWrapperNumber, aod, val : Variable, old : Variable ) : void {
            throw new Error( "TODO" );
        },
        update_patch_data_l0( rp: OtWrapperNumber, args ) : void {
        }
    };
    static op_add = {
        attrs: [ "Variable" ],
        right( rp: OtWrapperNumber, flags : number, nv : Variable ) : boolean {
            return nv.sup_immediate( 0 ) ? OtWrapperNumber.can_increase.ok( flags ) : nv.equ_immediate( 0 ) || OtWrapperNumber.can_decrease.ok( flags );
        },
        valid( rp: OtWrapperNumber, nv : Variable ) : boolean {
            throw new Error( "TODO" );
        },
        apply_op( rp: OtWrapperNumber, aod, nv : Variable ) : void {
            Method.methods["self_add_b"].call_2( rp.val, nv );
            rp.sig_change();
        },
        undo( rp: OtWrapperNumber, aod, nv : Variable ) : void {
            throw new Error( "TODO" );
        },
        update_patch_data_l0( rp: OtWrapperNumber, args ) : void {
        }
    };

    static op_types = OtWrapper.op_types.concat( [
        OtWrapperNumber.op_set,
        OtWrapperNumber.op_set_und,
        OtWrapperNumber.op_add
    ] );

    // right types
    static can_read = {
        name: "read",
        flags: 4,
        ok( flags : number ) : boolean { return ( flags & 4 ) != 0; }
    }
    static can_increase = {
        name: "increase",
        flags: 8,
        ok( flags : number ) : boolean { return ( flags & 8 ) != 0; }
    }
    static can_decrease = {
        name: "decrease",
        flags: 16,
        ok( flags : number ) : boolean { return ( flags & 16 ) != 0; }
    }

    static right_types = OtWrapper.right_types.concat( [
        OtWrapperNumber.can_read,
        OtWrapperNumber.can_increase,
        OtWrapperNumber.can_decrease
    ] );

    constructor( rp : Rp ) {
        super( OtWrapperNumber );
        this.val = new Number( rp );

        // first patch
        if ( ! this.val.equ_immediate( 0 ) )
            OtWrapperNumber.reg_op( this.last_ops, OtWrapperNumber.op_set, this.val );
    }

    prox_ref() : Variable {
        return this.val;
    }

    _self_set( nv : Variable, usr_id = this.usr_id ) {
        if ( ! nv.equ_immediate( this.val ) && OtWrapperNumber.op_set_und.right( this, this.usr_right( usr_id ), nv, this.val ) ) {
            OtWrapperNumber.reg_op( this.last_ops, OtWrapperNumber.op_set_und, nv, this.val );
            OtWrapperNumber.op_set_und.apply_op( this, { usr: usr_id, type: 0 }, nv, this.val );
            return true;
        }
        return false;
    }

    _self_add( nv : Variable, usr_id = this.usr_id ) {
        if ( ! nv.equ_immediate( 0 ) && OtWrapperNumber.op_add.right( this, this.usr_right( usr_id ), nv ) ) {
            OtWrapperNumber.reg_op( this.last_ops, OtWrapperNumber.op_add, nv );
            OtWrapperNumber.op_add.apply_op( this, { usr: usr_id, type: 0 }, nv );
            return true;
        }
        return false;
    }

    val : Variable;
}
OtWrapper.make_templ( OtWrapperNumber );


Method.methods["self_set_o"].add_surdef( 2, [ OtWrapperNumber, "to_Number" ], function( num : { rp: OtWrapperNumber }, sup : Variable ) {
    return num.rp._self_set( sup );
} );

Method.methods["self_set_b"].add_surdef( 2, [ OtWrapperNumber, "to_Number" ], function( num : { rp: OtWrapperNumber }, sup : Number ) {
    return num.rp._self_set( sup.copy() );
} );

Method.methods["self_add_b"].add_surdef( 2, [ OtWrapperNumber, "to_Number" ], function( num : { rp: OtWrapperNumber }, sup : Variable ) {
    return num.rp._self_add( sup );
} );

Method.methods["self_sub_b"].add_surdef( 2, [ OtWrapperNumber, "to_Number" ], function( num : { rp: OtWrapperNumber }, sup : Variable ) {
    return num.rp._self_add( Method.methods["neg_b"].call_1( sup ) );
} );

import Variable from "../Core/Variable";
import Rp       from "../Core/Rp";
import Number   from "../Number";
